routes/turnilo: allow query params on GET dashs

msune opened the pull request:
routes/turnilo: allow query params on GET dashs
    
Allow the following query parameters (exact match) on the main GET dashboards URL:
    
    * `shortName`
    * `dataCube`
    
This allows clients to validate whether a unique constrain `shortName+dataCube` is in use and by which dashboard ID.
    
Added test coverage too.

msune committed:
routes: move turnilo routes to dedicated router

This commit moves turnilo dashboard routes to a dedicated module.

msune committed:
routes/turnilo: allow query params on GET dashs

Allow the following query parameters (exact match) on the main
GET dashboards URL:

* `shortName`
* `dataCube`

This allows clients to validate whether a unique constrain
`shortName+dataCube` is in use and by which dashboard ID.

Added test coverage too.

msune committed:
requirements.txt: bump deps

Bump dependencies to latest, to fix an invalid Warning by sqlalchemy

msune committed:
services: fix leftovers alchemy->sqlmodel & warns

This commit:

* Cleans up some left overs from the transition of SQLAlchemy ->
  SQLModel
* Fixes annoying deprecation SQLModel warnings

msune committed:
test: turn warning as errors

Pass -Werror to pytest so that warnings become errors.

msune commented:
I think you will like what I piled on top @OscarMoya (yeah, I know, I should have stubbed this in another PR, but there were some conflicts):

```
commit 3505ac01085c44574dc2977a67dff3453b0af626 (HEAD -> turnilo_get_filter, origin/turnilo_get_filter)
Author: Marc Sune <marcdevel@gmail.com>
Date:   Wed Jun 12 23:48:23 2024 +0200

    services: fix leftovers alchemy->sqlmodel & warns
    
    This commit:
    
    * Cleans up some left overs from the transition of SQLAlchemy ->
      SQLModel
    * Fixes annoying deprecation SQLModel warnings

commit 086c861907c14bc72aaa6fd3bd548234537001a2
Author: Marc Sune <marcdevel@gmail.com>
Date:   Wed Jun 12 23:55:00 2024 +0200

    test: turn warning as errors
    
    Pass -Werror to pytest so that warnings become errors.

commit 0285f21652d9bd464aaae6c73f1791a42f492bc4
Author: Marc Sune <marcdevel@gmail.com>
Date:   Wed Jun 12 23:22:55 2024 +0200

    requirements.txt: bump deps
    
    Bump dependencies to latest, to fix an invalid Warning by sqlalchemy
```

OscarMoya commented on src/routes/turnilo_dashboard_routes.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def is_valid_query(s: str) -> bool:
+    if not isinstance(s, str):
```
After digging on the FastAPI docs, I think there is no need to check types, as they always be converted on the internal dependency chain and provided to the controller function or return queryError. 

Related to the length and pattern. FastApi let's you declare that in the query definition:
``` 
Query(          
            title="Query string",
            description="Query string for the items to search in the database that have a good match",
            min_length=3,
            max_length=50,
            pattern="^fixedquery$",
```
Although it is way more verbose, I personally don't like it, the advantage is that all of this will be added to the openAPI spec which will enrich the overall documentation of the API

OscarMoya replied:
Additionally, my preferred way would be to use pydantic models as the Query Drivers, They would be minimal, easy to document, typed, you can add additional validation, And you can use inheritance on them. Moreover, they look just a as single param in the functions ensuring minimal boilerplate and easy maintenance on the HTTP controllers, example: 

```python
from fastapi import FastAPI, Depends
from pydantic import BaseModel, Field
from typing import Optional

app = FastAPI()

class QueryParams(BaseModel):
    """
    Class representing the query params, there are 3 of them with types str, int and bool where
    the last one is optional
    """
    param1: str = Field(..., examples=["value1"], description="A required string parameter")
    param2: int = Field(..., examples=[123], description="A required integer parameter")
    param3: Optional[bool] = Field(None, examples=[True], description="An optional boolean parameter")

@app.get("/items/", summary="Get items", description="Retrieve items with the given query parameters")
async def read_items(query_params: QueryParams = Depends()):
    """
    This endpoint retrieves items based on the provided query parameters.
    - **param1**: A required string parameter
    - **param2**: A required integer parameter
    - **param3**: An optional boolean parameter
    """
    return {"param1": query_params.param1, "param2": query_params.param2, "param3": query_params.param3}

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="", port=8000)
```

after running this and calling it:

```
http://localhost:8000/items/?param1=exq&param2=123&param3=true&noexist=aaa
```
Mind the `noexist` param, you will get:

```json
{"param1":"exq","param2":123,"param3":true}
```
and of course, they also appear on the API documentation

![image](https://github.com/datahangar/backend/assets/1920980/ca56810d-ff82-44e9-b5b8-13e8f6a6b9e0)

OscarMoya replied:
The option I gave I think is way more extensible and easy to re-use and maintain, because if you ever need to put pagination in the routers, you will have to add and change all the signatures of the functions

msune replied:
Let's go with your wise advise:

1. Remove type validators
2. Go with the Query obj.

Give me some time to push it.

msune replied:
Applied; I couldn't find a way to have the `QueryParams` class in the routes module, instead of the services, due to a circulate dependency.

I think it's kind of unavoidable, due to typing... I cannot import it in the function signature. But if you know how to do it, lemme know :rocket:

OscarMoya replied:
I don't think is a big deal and I liked were thinks stand, as a workaround you can annotate the function with "" instead of the class type. But is usually advisable not to have circular dependencies even if it is for annotation purtposes

OscarMoya commented on src/routes/turnilo_dashboard_routes.py:
```diff
@@ -0,0 +1,85 @@
+        raise HTTPException(status_code=400, detail=f"Invalid shortName='{shortName}'")
+    if dataCube and not is_valid_query(dataCube):
+        raise HTTPException(status_code=400, detail=f"Invalid dataCube='{dataCube}'")
+    return td.dashboards_get_all(db_session, shortName, dataCube)
```
Even if you decide to not to go to the BaseModel approach for query params, it would be great if you protect your DB functions from them. I would aggregate in some kind of Filter function like this: 

```python
f := DashFilter(data_cube, name)
td.dashboards_get_all(db_session, f) # this dashFilter can dump easily filter dicts and if more Queries appear you do not need to refactor all your libraries to accomodate new fields
```

msune replied:
N/A, will follow your approach above. Closing.

OscarMoya reviewed (commented):
If I were you, I would consider adopt some of my proposals to make sure the code is secure and extensible with minimal boilerplate, but as it is right now it would work

msune committed:
requirements.txt: move to psycopg2-binary

Make it easy to run within a venv.

msune committed:
svcs|routes.turnilo_dashboards: adopt Query class

Adopt Oscar's suggestion to use a Query class to validate query
parameters.

Unfortunately, due to the circular dependency, the Query class
is moved to services (due to typing).

Signed-off-by: Marc Sune <marcdevel@gmail.com>
Signed-off-by: Oscar Moya <o.moya.gomez@gmail.com>

OscarMoya reviewed (approved):
LGTM
